File: app/src/main/java/com/stevenfrew/beatprompter/ChooseSongMessage.java
package com.stevenfrew.beatprompter;

import android.util.Log;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;

class ChooseSongMessage extends BluetoothMessage
{
    static final byte CHOOSE_SONG_MESSAGE_ID=0;

    String mTitle;
    String mTrack;
    boolean mBeatScroll;
    boolean mSmoothScroll;
    int mOrientation;
    int mMinFontSize;
    int mMaxFontSize;
    int mScreenWidth;
    int mScreenHeight;

    ChooseSongMessage(LoadingSongFile lsf)
    {
        mTitle=lsf.mSongFile.mTitle;
        if(mTitle==null)
            mTitle="";
        mTrack=lsf.mTrack;
        mOrientation=lsf.mNativeDisplaySettings.mOrientation;
        if(mTrack==null)
            mTrack="";
        mBeatScroll=lsf.mScrollMode==ScrollingMode.Beat;
        mSmoothScroll=lsf.mScrollMode==ScrollingMode.Smooth;
        mMinFontSize=lsf.mNativeDisplaySettings.mMinFontSize;
        mMaxFontSize=lsf.mNativeDisplaySettings.mMaxFontSize;
        mScreenWidth=lsf.mNativeDisplaySettings.mScreenWidth;
        mScreenHeight=lsf.mNativeDisplaySettings.mScreenHeight;
    }

    ChooseSongMessage(byte[] bytes) throws NotEnoughBluetoothDataException
    {
        try
        {
            final ByteArrayInputStream byteArrayInputStream =
                    new ByteArrayInputStream(bytes);
            int dataRead=byteArrayInputStream.read(new byte[1]);
            if(dataRead==1) {
                int availableStart=byteArrayInputStream.available();
                final ObjectInputStream objectInputStream =
                        new ObjectInputStream(byteArrayInputStream);
                mTitle = (String) objectInputStream.readObject();
                mTrack = (String) objectInputStream.readObject();
                mBeatScroll = objectInputStream.readBoolean();
                mSmoothScroll = objectInputStream.readBoolean();
                mOrientation = objectInputStream.readInt();
                mMinFontSize = objectInputStream.readInt();
                mMaxFontSize = objectInputStream.readInt();
                mScreenWidth=objectInputStream.readInt();
                mScreenHeight=objectInputStream.readInt();
                int availableEnd=byteArrayInputStream.available();
                objectInputStream.close();
                byteArrayInputStream.close();
                mMessageLength=1+(availableStart-availableEnd);
            }
            else
                throw new NotEnoughBluetoothDataException();
        }
        catch(Exception e)
        {
            Log.e(SongList.TAG,"Couldn't read ChooseSongMessage data, assuming not enough data",e);
            throw new NotEnoughBluetoothDataException();
        }
    }

    public byte[] getBytes()
    {
        try
        {
            final ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
            byteArrayOutputStream.write(new byte[]{CHOOSE_SONG_MESSAGE_ID},0,1);
            final ObjectOutputStream objectOutputStream =
                    new ObjectOutputStream(byteArrayOutputStream);
            objectOutputStream.writeObject(mTitle);
            objectOutputStream.writeObject(mTrack);
            objectOutputStream.writeBoolean(mBeatScroll);
            objectOutputStream.writeBoolean(mSmoothScroll);
            objectOutputStream.writeInt(mOrientation);
            objectOutputStream.writeInt(mMinFontSize);
            objectOutputStream.writeInt(mMaxFontSize);
            objectOutputStream.writeInt(mScreenWidth);
            objectOutputStream.writeInt(mScreenHeight);
            objectOutputStream.flush();
            objectOutputStream.close();

            final byte[] byteArray = byteArrayOutputStream.toByteArray();
            byteArrayOutputStream.close();

            return byteArray;
        }
        catch(Exception e)
        {
            Log.e(SongList.TAG,"Couldn't write ChooseSongMessage data.",e);
        }
        return null;
    }
}

File: app/src/main/kotlin/com/stevenfrew/beatprompter/comm/bluetooth/message/ChooseSongMessage.kt
package com.stevenfrew.beatprompter.comm.bluetooth.message

import com.stevenfrew.beatprompter.Logger
import com.stevenfrew.beatprompter.graphics.Rect
import com.stevenfrew.beatprompter.song.load.SongChoiceInfo
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream

/**
 * OutgoingMessage that is sent/received when a song is chosen.
 */
class ChooseSongMessage(
	bytes: ByteArray,
	val choiceInfo: SongChoiceInfo
) : BluetoothMessage(bytes) {
	constructor(choiceInfo: SongChoiceInfo) : this(asBytes(choiceInfo), choiceInfo)

	companion object {
		private fun asBytes(choiceInfo: SongChoiceInfo): ByteArray =
			ByteArrayOutputStream().apply {
				write(byteArrayOf(CHOOSE_SONG_MESSAGE_ID), 0, 1)
				ObjectOutputStream(this).apply {
					writeObject(choiceInfo.normalizedTitle)
					writeObject(choiceInfo.normalizedArtist)
					writeObject(choiceInfo.variation)
					writeBoolean(choiceInfo.isBeatScroll)
					writeBoolean(choiceInfo.isSmoothScroll)
					writeInt(choiceInfo.orientation)
					writeFloat(choiceInfo.minFontSize)
					writeFloat(choiceInfo.maxFontSize)
					writeInt(choiceInfo.screenSize.width)
					writeInt(choiceInfo.screenSize.height)
					writeBoolean(choiceInfo.noAudio)
					writeInt(choiceInfo.audioLatency)
					writeInt(choiceInfo.transposeShift)
					flush()
					close()
				}
			}.toByteArray()

		internal fun fromBytes(bytes: ByteArray): ChooseSongMessage {
			try {
				ByteArrayInputStream(bytes).apply {
					val dataRead = read(ByteArray(1))
					if (dataRead == 1) {
						val availableStart = available()
						val songChoiceInfo =
							ObjectInputStream(this).run {
								val title = readObject() as String
								val artist = readObject() as String
								val variation = readObject() as String
								val beatScroll = readBoolean()
								val smoothScroll = readBoolean()
								val orientation = readInt()
								val minFontSize = readFloat()
								val maxFontSize = readFloat()
								val screenWidth = readInt()
								val screenHeight = readInt()
								val noAudio = readBoolean()
								var audioLatency = 0
								var transposeShift = 0
								try {
									audioLatency = readInt()
									transposeShift = readInt()
								} catch (_: Exception) {
									// Old versions will not send these last two items of data.
									// Try to cope.
								}
								SongChoiceInfo(
									title,
									artist,
									variation,
									orientation,
									beatScroll,
									smoothScroll,
									minFontSize,
									maxFontSize,
									Rect(0, 0, screenWidth, screenHeight),
									noAudio,
									audioLatency,
									transposeShift
								)
							}
						val availableEnd = available()
						val messageLength = 1 + (availableStart - availableEnd)
						close()
						Logger.logLoader({ "Received Bluetooth request to load \"${songChoiceInfo.normalizedTitle}\"" })
						return ChooseSongMessage(bytes.copyOfRange(0, messageLength), songChoiceInfo)
					}
				}
			} catch (e: Exception) {
				Logger.logComms({ "Couldn't read ChooseSongMessage data, assuming not enough data" }, e)
			}
			throw NotEnoughDataException()
		}
	}
}
